logictest: protected timestamp record does not exist

pkg/sql/logictest/tests/fakedist/fakedist_test.TestLogic_alter_primary_key [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7691493?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7691493?buildTab=artifacts#/) on master @ [b5be006bedd7d3cedc3fb3d2248df168e3d64be2](https://github.com/cockroachdb/cockroach/commits/b5be006bedd7d3cedc3fb3d2248df168e3d64be2):


```
=== RUN   TestLogic_alter_primary_key
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/d549810e45034f43a9494d02b085e623/logTestLogic_alter_primary_key1996763357
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestLogic_alter_primary_key
    logic.go:4129: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/d549810e45034f43a9494d02b085e623/logTestLogic_alter_primary_key1996763357
--- FAIL: TestLogic_alter_primary_key (7.53s)
=== RUN   TestLogic_alter_primary_key/index_rewrites
    logic.go:3182: 
        
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/2386/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/fakedist/fakedist_test_/fakedist_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/alter_primary_key:258: 
        expected success, but found
        (XXUUU) job 816917427264749569: protected timestamp record does not exist
        protectedts.go:31: in init()
    logic.go:2144: 
         pq: job 816917427264749569: protected timestamp record does not exist
[10:56:54] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/2386/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/fakedist/fakedist_test_/fakedist_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/alter_primary_key with config fakedist: 60 tests, 2 failures
[10:56:55] --- total progress: 60 statements
--- total: 60 tests, 2 failures
    --- FAIL: TestLogic_alter_primary_key/index_rewrites (0.92s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_alter_primary_key.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-21810

--- a/pkg/jobs/jobsprotectedts/jobs_protected_ts_manager.go
+++ b/pkg/jobs/jobsprotectedts/jobs_protected_ts_manager.go
@@ -192,6 +192,7 @@ func (p *Manager) Protect(
 		return nil, err
 	}
 	return func(ctx context.Context) error {
+		// Remove the protected timestamp.
 		return p.Unprotect(ctx, job)
 	}, nil
 }
@@ -201,19 +202,23 @@ func (p *Manager) Protect(
 // record. Note: This should only be used for job cleanup if is not currently,
 // executing.
 func (p *Manager) Unprotect(ctx context.Context, job *jobs.Job) error {
-	return p.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
-		// Fetch the protected timestamp UUID from the job, if one exists.
-		protectedtsID := getProtectedTSOnJob(job.Details())
+	// Fetch the protected timestamp UUID from the job, if one exists.
+	if getProtectedTSOnJob(job.Details()) == nil {
+		return nil
+	}
+	// If we do find one then we need to clean up the protected timestamp,
+	// and remove it from the job.
+	return job.Update(ctx, nil, func(txn *kv.Txn, md jobs.JobMetadata, ju *jobs.JobUpdater) error {
+		// The job will get refreshed, so check one more time the protected
+		// timestamp still exists. The callback returned from Protect works
+		// on a previously cached copy.
+		protectedtsID := getProtectedTSOnJob(md.Payload.UnwrapDetails())
 		if protectedtsID == nil {
 			return nil
 		}
-		// If we do find one then we need to clean up the protected timestamp,
-		// and remove it from the job.
-		return job.Update(ctx, txn, func(txn *kv.Txn, md jobs.JobMetadata, ju *jobs.JobUpdater) error {
-			updatedDetails := setProtectedTSOnJob(job.Details(), nil)
-			md.Payload.Details = jobspb.WrapPayloadDetails(updatedDetails)
-			ju.UpdatePayload(md.Payload)
-			return p.protectedTSProvider.Release(ctx, txn, *protectedtsID)
-		})
+		updatedDetails := setProtectedTSOnJob(job.Details(), nil)
+		md.Payload.Details = jobspb.WrapPayloadDetails(updatedDetails)
+		ju.UpdatePayload(md.Payload)
+		return p.protectedTSProvider.Release(ctx, txn, *protectedtsID)
 	})
 }

